File: kashmir-backend/app/services/payments.py
import hashlib
import time
from datetime import datetime, timedelta, timezone
from jose import jwt, JWTError
from app.config import get_settings
from app.models.schemas import (
    AirpayCreateRequest,
    AirpayCreateResponse,
    PaymentVerifyResponse,
)


def _clean(val: str) -> str:
    return val.strip().strip("'\"")


def create_airpay_order(req: AirpayCreateRequest) -> AirpayCreateResponse:
    s = get_settings()

    merchant_id = _clean(s.airpay_merchant_id)
    username = _clean(s.airpay_username)
    password = _clean(s.airpay_password)
    api_key = _clean(s.airpay_api_key)

    txn_id = f"KFP{int(time.time())}"
    amount = str(s.documentary_price_inr)
    currency = "356"  # INR

    private_key = hashlib.sha256(
        f"{api_key}@{username}:|:{password}".encode()
    ).hexdigest()

    s_key = hashlib.sha256(
        f"{username}~:~{password}".encode()
    ).hexdigest()

    buyer_data = "~:~".join([
        req.email, req.name, req.name, "",
        req.address, req.city, req.state, req.country, amount,
        txn_id, currency, "0", "INR", "0",
        req.phone, req.pin_code,
    ])
    checksum = hashlib.sha256(
        f"{s_key}@{buyer_data}".encode()
    ).hexdigest()

    form_fields = {
        "mercid": merchant_id,
        "orderid": txn_id,
        "amount": amount,
        "currency": currency,
        "isocurrency": "INR",
        "chmod": "0",
        "buyerEmail": req.email,
        "buyerPhone": req.phone,
        "buyerFirstName": req.name,
        "buyerLastName": "",
        "buyerAddress": req.address,
        "buyerCity": req.city,
        "buyerState": req.state,
        "buyerCountry": req.country,
        "buyerPinCode": req.pin_code,
        "privatekey": private_key,
        "checksum": checksum,
        "txnsubtype": "",
    }

    return AirpayCreateResponse(
        transaction_id=txn_id,
        post_url=s.airpay_base_url,
        form_fields=form_fields,
    )


def verify_airpay_callback(form_data: dict) -> PaymentVerifyResponse:
    s = get_settings()
    secret = _clean(s.airpay_secret_key)

    txn_id = form_data.get("TRANSACTIONID", "")
    ap_txn_id = form_data.get("APTRANSACTIONID", "")
    amount = form_data.get("AMOUNT", "")
    status = form_data.get("TRANSACTIONSTATUS", "")
    message = form_data.get("MESSAGE", "")
    ap_hash = form_data.get("ap_SecureHash", "")

    verify_str = f"{txn_id}:{ap_txn_id}:{amount}:{status}:{message}:{secret}"
    expected = hashlib.sha256(verify_str.encode()).hexdigest()

    if expected != ap_hash:
        return PaymentVerifyResponse(verified=False, message="Invalid checksum")

    if str(status) != "200":
        return PaymentVerifyResponse(verified=False, message=f"Payment failed: {message}")

    expire = datetime.now(timezone.utc) + timedelta(minutes=s.access_token_expire_minutes)
    token = jwt.encode(
        {"sub": ap_txn_id or txn_id, "exp": expire},
        s.jwt_secret,
        algorithm=s.jwt_algorithm,
    )
    return PaymentVerifyResponse(verified=True, access_token=token, message="Payment verified")


def verify_access_token(token: str) -> dict | None:
    s = get_settings()
    try:
        return jwt.decode(token, s.jwt_secret, algorithms=[s.jwt_algorithm])
    except JWTError:
        return None

File: kashmir-frontend/src/server/payments.ts
import crypto from 'crypto';
import { SignJWT, jwtVerify, type JWTPayload } from 'jose';
import { getServerSettings } from '@/server/config';

export interface AirpayOrderInput {
  email: string;
  name: string;
  phone: string;
  address?: string;
  city?: string;
  state?: string;
  country?: string;
  pin_code?: string;
}

export interface AirpayOrderResult {
  transaction_id: string;
  post_url: string;
  form_fields: Record<string, string>;
}

export interface VerifyCallbackResult {
  verified: boolean;
  access_token?: string;
  message: string;
}

function clean(val: string): string {
  return val.trim().replace(/^['"]|['"]$/g, '');
}

function sha256(input: string): string {
  return crypto.createHash('sha256').update(input).digest('hex');
}

export async function createAirpayOrder(input: AirpayOrderInput): Promise<AirpayOrderResult> {
  const s = getServerSettings();

  const merchantId = clean(s.airpayMerchantId);
  const username = clean(s.airpayUsername);
  const password = clean(s.airpayPassword);
  const apiKey = clean(s.airpayApiKey);

  const txnId = `KFP${Date.now()}`;
  const amount = String(s.documentaryPriceInr);
  const currency = '356'; // INR

  const privateKey = sha256(`${apiKey}@${username}:|:${password}`);
  const sKey = sha256(`${username}~:~${password}`);

  const buyerData = [
    input.email, input.name, input.name, '',
    input.address ?? '', input.city ?? '', input.state ?? '',
    input.country ?? 'India', amount, txnId, currency, '0', 'INR', '0',
    input.phone, input.pin_code ?? '',
  ].join('~:~');

  const checksum = sha256(`${sKey}@${buyerData}`);

  const formFields: Record<string, string> = {
    mercid: merchantId,
    orderid: txnId,
    amount,
    currency,
    isocurrency: 'INR',
    chmod: '0',
    buyerEmail: input.email,
    buyerPhone: input.phone,
    buyerFirstName: input.name,
    buyerLastName: '',
    buyerAddress: input.address ?? '',
    buyerCity: input.city ?? '',
    buyerState: input.state ?? '',
    buyerCountry: input.country ?? 'India',
    buyerPinCode: input.pin_code ?? '',
    privatekey: privateKey,
    checksum,
    txnsubtype: '',
  };

  return {
    transaction_id: txnId,
    post_url: s.airpayBaseUrl,
    form_fields: formFields,
  };
}

export async function verifyAirpayCallback(formData: Record<string, string>): Promise<VerifyCallbackResult> {
  const s = getServerSettings();
  const secret = clean(s.airpaySecretKey);

  const txnId = formData.TRANSACTIONID ?? '';
  const apTxnId = formData.APTRANSACTIONID ?? '';
  const amount = formData.AMOUNT ?? '';
  const status = formData.TRANSACTIONSTATUS ?? '';
  const message = formData.MESSAGE ?? '';
  const apHash = formData.ap_SecureHash ?? '';

  const verifyStr = `${txnId}:${apTxnId}:${amount}:${status}:${message}:${secret}`;
  const expected = sha256(verifyStr);

  if (expected !== apHash) {
    return { verified: false, message: 'Invalid checksum' };
  }

  if (status !== '200') {
    return { verified: false, message: `Payment failed: ${message}` };
  }

  const key = new TextEncoder().encode(s.jwtSecret);
  const expSeconds = Math.floor(Date.now() / 1000) + s.accessTokenExpireMinutes * 60;
  const token = await new SignJWT({})
    .setProtectedHeader({ alg: s.jwtAlgorithm })
    .setSubject(apTxnId || txnId)
    .setExpirationTime(expSeconds)
    .sign(key);

  return { verified: true, access_token: token, message: 'Payment verified' };
}

export async function verifyAccessToken(token: string): Promise<JWTPayload | null> {
  const s = getServerSettings();
  try {
    const key = new TextEncoder().encode(s.jwtSecret);
    const { payload } = await jwtVerify(token, key, { algorithms: [s.jwtAlgorithm] });
    return payload;
  } catch {
    return null;
  }
}
